lib/dune_file.ml

```diff
@@ -3,39 +3,77 @@ open Import
 module Lang = struct
   type version = int * int
 
-  let compare_version (major, minor) (major', minor') =
-    match Int.compare major major' with
-    | 0 -> Int.compare minor minor'
-    | _ as ord -> ord
-
   let pp_version fmt (major, minor) = Format.fprintf fmt "%d.%d" major minor
 
-  let parse_version s =
-    let err () =
-      Error (`Msg (Printf.sprintf "Invalid dune lang version: %s" s))
-    in
-    match String.split_on_char ~sep:'.' s with
-    | [ major; minor ] -> (
-        match (int_of_string_opt major, int_of_string_opt minor) with
-        | Some major, Some minor -> Ok (major, minor)
-        | _ -> err ())
-    | _ -> err ()
-
-  let parse_stanza s =
-    let content =
-      let open Option.O in
-      String.drop_prefix ~prefix:"(" s >>= String.drop_suffix ~suffix:")"
-      >>| fun content -> String.split_on_char ~sep:' ' content
-    in
-    match content with
-    | Some [ "lang"; "dune"; version ] -> parse_version version
-    | _ -> Error (`Msg (Printf.sprintf "Invalid lang stanza: %s" s))
+  let version_to_string version = Format.asprintf "%a" pp_version version
 
-  let is_stanza s = String.is_prefix ~prefix:"(lang " s
+  let stanza_regexp =
+    let open Re in
+    seq
+      [
+        bol;
```

Comment:
the lang stanza is only valid as first line, so if you want you can change that to `bos` and remove the error case corresponding to multiple stanzas below

Reply:
Can you confirm this has always been the case?

Reply:
One thing worth noting is that if we do this, we won't be able to detect bad dune-projects anymore.

I think we can safely only accept them, and simply warn that the dune-project is invalid if we can't find the lang stanza (instead of prepending it).

Reply:
> Can you confirm this has always been the case?

I think that this forward compatibility has been considered since the release of dune 1.0.0.